### ISSUE TITLE
<chart>: standard chart types (bar, line, scatter, pie, box, histogram)

### ISSUE THREAD

**Issue by virginiaMae** [Doenet/DoenetML#437]
Originally filed as a request for pie, bar and line charts. This is now the umbrella/tracking issue for the whole of `<chart>`: it records the current state, the design decisions, and the phased roadmap. Each phase has its own linked issue.

## Roadmap checklist

- [x] #1833 — `<chart type="bar">`, rendered with PreFigure (shipped in #1829, renamed from `<barChart>` in #1869)
- [x] #1877 — Phase 1: `<series>`, legend, title, and grouped/stacked bars (PR #1883)
- [x] #1878 — Phase 2: a numeric x-axis, `type="line"` and `type="scatter"` (PR #1884, stacked on #1883)
- [x] #1879 — Phase 3: `type="pie"`
- [ ] #1880 — Phase 4: `type="box"`
- [ ] #1881 — Phase 5: `type="histogram"`
- [ ] #1882 — Docs: a reference section per chart type (the restructure and the guide split landed in #1883)

Ordering rationale: Phase 1 is the interface work every later type depends on, and it lands with the type we already ship, so the shape can be checked against something known-good before four more types are built on it. Phase 2 pairs line and scatter because both need the same thing — a numeric horizontal axis, which the categorical `<tick-mark>` axis cannot express — and nothing else does. Phases 3–5 are then independent of each other and can land in any order or in parallel; pie is listed first because it is the type most often asked for after bar. Each phase adds its own section to the reference page; the docs issue is the restructure those sections need to slot into, and should land with or immediately after Phase 1.

## Current state

`<chart type="bar">` ships today: `components/Chart.js`, `utils/prefigure/chart.ts`, `Viewer/renderers/chart.tsx`, `pages/reference/chart.mdx`. Children are the values, `categories` names the bars, and the geometry lives in a renderer-neutral `chartGeometry` state variable that `prefigureXML` serializes, so a chart's meaning does not depend on PreFigure drawing it.

`type` is required and has no default, and it is checked in exactly one place — `chartGeometry` returns null for a type the component cannot draw, and everything downstream reads the absence from there. Each new type is therefore a branch in that one definition plus a builder in `utils/prefigure/`.

## What PreFigure can actually draw

Only two chart tags are built in — `<scatter>` and `<histogram>` (`prefig/core/statistics.py`). There is no bar, pie, line or box tag: `type="bar"` is assembled from `<rectangle>` and `<tick-mark>`. So "if PreFigure has it built in" is not a usable test — it would exclude the type we already ship. The test that matters is whether a type can be assembled from PreFigure's primitives and stay navigable in diagcess:

| type | drawn with |
| --- | --- |
| bar | `<rectangle>` (already) |
| line | `<polygon closed="no">` |
| scatter | `<scatter>`, or `<point>` per datum for per-point annotations |
| pie | `<arc sector="yes">` |
| box | `<rectangle>` + `<line>` + `<point>` |
| histogram | `<rectangle>`, binned in the worker |

Deliberately out of scope: **area** (a closed polyline — nearly free once line exists, but nothing depends on it), **dot/strip plots**, **violin** (needs a KDE), and stem-and-leaf (not a drawing). Revisit area and dot plots once the six above have landed.

`<histogram>` is not used even though it exists, for the reason #1833 gave: it bins in scipy and emits rectangles, which leaves the counts unreachable from `$…` for a sentence, a table or an `<answer>`. Binning in the worker keeps them readable, and shares its definition with `<binCounts>`.

### One tag with a `type`, not a tag per chart

Already settled by #1869. A pie chart and a box plot differ in how the same data is drawn, not in what an author is doing, and `type` is an attribute so it can be computed — a document can chart the same data both ways without duplicating the tag around it.

### Multiple series: a `<series>` child

`values` is flat today, and the types differ in what data they need — one value per category (bar, line, pie), (x, y) pairs (scatter), raw observations (box, histogram). Every standard statistical plotting package solves this the same way: data, a mark, and encodings that map fields onto x, y and color. The markup equivalent is a `<series>` child that carries one group of data plus its own legend label and style.

```
<chart type="bar" categories="North South East West" layout="grouped">
  <title>Population by region</title>
  <yLabel>people</yLabel>
  <series label="2024">41 63 18 78</series>
  <series label="2025" styleNumber="2">45 60 22 80</series>
</chart>
```

That maps onto a Vega-Lite-style spec without a rewrite later: `type` is the mark, a `<series>` is the color/group channel, its children and `x` are the y and x channels, and `categories`/`yMin`/`yMax` are scale domains. It also gives diagcess a real tree to walk — figure, then series, then mark — which a flat list of values cannot.

A chart with bare numbers and no `<series>` is one unnamed series, which is exactly what `<chart>` does today, so nothing already written changes.

### Title and legend

Both are drawn by PreFigure rather than by the renderer, so they survive SVG export, print and tactile output.

- **Title** — a `<title>` child, drawn as a `<label>` anchored at the top of the box with the top margin widened to hold it. PreFigure's own `<caption>` only affects tactile output, so it is set alongside the label rather than instead of it.
- **Legend** — PreFigure's `<legend>` with one `<item ref="…">` per series, keyed off the `at` handles the marks already carry. Two constraints, both handled in Phase 1: it hardcodes `fill="white"` on its background box, so `opacity="0"` plus `stroke="currentColor"` is what makes it readable in dark mode until a `fill` attribute can be contributed upstream; and it measures label text through the browser canvas in Pyodide (`PyodideTextMeasurements`), which works but has to be confirmed against a real render.

### Colors stay semantic

A series is colored by `styleNumber`, like every other Doenet component — not by a color attribute. See `concepts/styling`.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01RqRJ3QoH4UrHFR41aAkN8e

**Comment by dqnykamp** [comment]
The **bar** third of this is now in flight: #1833, PR #1829. It renders through the PreFigure pipeline rather than a JS charting library, which gets a categorical axis (`<graph>` has no custom-tick support at all) and diagcess screen-reader navigation of the chart, and adds no dependency to the build — the Plotly cost that got the old `<chart>` component's renderer commented out in `0e4fe1ded` is not paid, because the PreFigure runtime is fetched lazily and only by documents that contain a diagram.

Pie and line charts are not addressed, so this issue stays open for them.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_016YdEWccVoUPCJoDYxKmHRo
